packages/okta-react/src/Security.js

```diff
@@ -10,34 +10,35 @@
  * See the License for the specific language governing permissions and limitations under the License.
  */
 
-import React, { Component } from 'react';
-import { withRouter } from 'react-router';
-import PropTypes from 'prop-types';
+import React, { useState, useEffect } from 'react';
 import Auth from './Auth';
+import OktaContext from './OktaContext';
 
-class Security extends Component {
-  constructor(props) {
-    super(props);
-    this.auth = props.auth || new Auth(props);
-  }
+const Security = (props) => { 
 
-  getChildContext() {
-    return {
-      auth: this.auth
-    };
-  }
+  const [auth] = useState( props.auth || new Auth(props) );
+  const [authState, setAuthState] = useState({...Auth.DEFAULT_STATE});
 
-  static childContextTypes = {
-    auth: PropTypes.object.isRequired
-  }
+  console.log('in security render');
 
-  render() {
-    return (
-      <div className={this.props.className}>
-        {this.props.children}
+  useEffect( () => { 
+    let unsub;
+    unsub = auth.on('authStateChange', (newAuthState) => { 
```

Comment:
How about leverage the state management practice introduced from react? Current implementation looks like states are coupled with `Auth` class. However, when reading through `Auth` class, seems like it's only a wrapper of auth.js, which probably can be extract out of this react repo in the future.

Reply:
Can you clarify what that would look like?  
- We need it to trigger on the initial render
- We need it to update after an async process
- We need to make sure later changes in state DON'T trigger a recheck of the auth

Reply:
My idea is we can leverage `context` and `useReducer` to store/update state globally. I learned the pattern from https://medium.com/simply/state-management-with-react-hooks-and-context-api-at-10-lines-of-code-baf6be8302c
It works like `redux`, but without `redux` dependency.

Reply:
I'm still not quite following.  That article uses `useState`, `useReducer` and a context.  

I'm using `useState`, `useEffect`, and a context.  

Even if I used `useReducer`, I'd still have to subscribe to the event emitter, so I'd still have to have `useEffect`.  It looks like what you're suggesting simply adds a layer over being able to update state when the status changes.  What am I missing?

Reply:
My suggestion is put `useReducer` in `Provider`, then you can have access to the global state (authState) and the `dispatch` function to make state change from context. Then by following the `flux` flow (action -> reducer) should be able to make state change properly. I am not against the `observer pattern` introduced in this PR, just because this pr is targeting to specific framework, it might be better to follow the popular pattern in the framework's ecosystem.

Reply:
My concern is that we'd be adding an abstraction layer to cover just one case - where we use the dispatcher in a callback to this event emitter.

Not having a reducer at all and just updating state is also a popular pattern, used when you don't have a complex state that involves multiple actions.  

Why do you think this state management is complex enough to justify a reducer?

Reply:
By browsing through all three packages for `react`, 'angular' and `vue`. My feeling is we probably can extract the `Auth.js` out from framework repos, since they are all doing basically same work. If that can be the case in the future, I would suggest not couple state management with `Auth.js`.

Reply:
Careful: the devil is in the details.  Auth.js translates between authJS and the application, and angular, vue, and react have some different expectations.

The current React SDK has no state management in AuthJS, and it is a nightmare of async calls that is a hotbed of race conditions.

Without being too abstract, what are the exact lines you see separating a unified Auth.js from the three SDKs?  (and yet, separating it from the truly generic authjs?)

Reply:
Agree, over abstract across repos without careful design can be a maintenance nightmare. I may revisit this idea in the future.

Reply:
@swiftone  & I did have some discussions about having the state separated from the service, following a "store/emitter" pattern for the state. The article @shuowu-okta  linked to has the same idea, but adds the "reducer" concept (which I think is probably not necessary) and uses the context as the store (which is a minor React-specific optimization). We *could* put the current logic into a barebones "AuthStateProvider" or "AuthStateStore" which holds the state, leaving the Auth wrapper stateless.

I think that ultimately we would like to normalize the services and push the implementation down a layer to `okta-auth-js` so that these SDK's can be simplified. Our users could refer to a single set of documentation for configuration and API methods in `okta-auth-js`, and the individual SDK's would not be hiding anything via a containing wrapper.

Incidentally, the underlying service [DOES have an emitter](https://github.com/okta/okta-auth-js/blob/master/packages/okta-auth-js/lib/browser/browser.js#L160), used by the TokenManager. I was *this* close to adding an `on` method on the core service, but I didn't have a specific event to expose. I actually think this `authStateChange` event  is a GREAT candidate for something we can push down into the core service. It's a VERY useful addition. Why should it only be for React? 

After this release we could take this functionality and plug it in directly to `okta-auth-js`, it would not change the contract we are putting forward here. You would still use `auth.on(`authStateChange`...`